lib/bridge.js

```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+    const dataReadyLimited = this._limited(dataReady, request);
 
-    if (this.opts.queue.type === "none") { // consume as soon as we have context
-        promise.done(() => {
-            this._onConsume(null, data);
-        }, (err) => {
-            this._onConsume(err);
-        });
-        return request.getPromise();
-    }
+    this._queue.push(event, dataReadyLimited);
+    this._queue.consume();
+
+    return (
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
 
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
+ */
+Bridge.prototype._limited = function(promise, request) {
+    // queue.perRequest controls whether multiple request can be processed by
+    // the bridge at once.
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        const promiseLimited = this._prevRequestPromise.reflect().return(promise);
         this._prevRequestPromise = request.getPromise();
+        return promiseLimited;
     }
 
-    this._queue.push(event, data, promise);
-    this._queue.consume();
-    return request.getPromise();
-};
+    return promise;
+}
 
 Bridge.prototype._onConsume = function(err, data) {
-    if (!err) {
-        this.opts.controller.onEvent(data.request, data.context);
+    if (err) {
+        // The data for the event could not be retrieved.
+        // In this case we reject the request and the event will be delivered
+        // by the HS to the bridge again.
+        data.request.reject();
+        this.opts.controller.onLog("onEvent failure: " + err);
         return;
     }
-    if (!this.opts.controller.onLog) {
-        return;
+
+    this.opts.controller.onEvent(data.request, data.context);
+};
+
+Bridge.prototype._getBridgeContext = function(event) {
+    if (this.opts.disableContext) {
+        return Promise.resolve(null);
     }
-    this.opts.controller.onLog(
-        "onEvent failure: " + err
+
+    const context = new BridgeContext({
+        sender: event.user_id,
+        target: event.state_key,
+        room: event.room_id
+    });
+    const contextReady = context.get(this._roomStore, this._userStore)
+
+    return contextReady.return(context);
+}
+
+Bridge.prototype._handleEventError = function(event, error) {
+    if (!error instanceof EventNotHandledError) {
+        error = wrap(error, InternalError);
+    }
+    this._botIntent.signalBridgeError(
+        event.room_id,
+        event.event_id,
+        "Discord",
+        error.reason,
+        this._getUserRegex()
     );
 };
 
+/**
+ * Returns a regex matching all users of the bridge.
+ *
+ * @return {string} Super regex composed of all user regexes.
+ */
+Bridge.prototype._getUserRegex = function() {
```

Comment:
I'd expect a RegExp object here :)

Reply:
You suggest changing the name to something like `_getUserRegexString`?

Reply:
I mean that this function should return a RegExp object, not a string.

Reply:
Just to turn it back into a string? O_o The method is only there to get the string representation of the regex for the event.

Reply:
My bad, you're not using it as a regex anywhere here but just to stick inside the event. Got it.